internal/controllers/machinedeployment/machinedeployment_rollout_planner_test.go

```diff
@@ -535,15 +598,93 @@ func machineSetControllerMutator(log *fileLogger, ms *clusterv1.MachineSet, scop
 		}
 	}
 
-	// if too many replicas, delete exceeding machines.
-	// exceeding machines are deleted in predictable order, so it is easier to write test case and validate rollout sequences.
+	// if too many replicas, delete or move exceeding machines.
+	// exceeding machines are deleted or moved in predictable order, so it is easier to write test case and validate rollout sequences.
 	// e.g. if a ms has m1,m2,m3 created in this order, m1 will be deleted first, then m2 and finally m3.
-	machinesToDelete := max(ptr.Deref(ms.Status.Replicas, 0)-ptr.Deref(ms.Spec.Replicas, 0), 0)
+	// Note: replicas still pending AcknowledgeMove should not be counted when computing the numbers of machines to delete, because those machines are not included in ms.Spec.Replicas yet.
+	// Without this check, the following logic would try to align the number of replicas to "an incomplete" ms.Spec.Replicas thus wrongly deleting replicas that should be preserved.
+	machinesToDeleteOrMove := max(ptr.Deref(ms.Status.Replicas, 0)-int32(notAcknowledgeMoveReplicas.Len())-ptr.Deref(ms.Spec.Replicas, 0), 0)
+	if machinesToDeleteOrMove > 0 {
+		if targetMSName, ok := ms.Annotations[clusterv1.MachineSetMoveMachinesToMachineSetAnnotation]; ok && targetMSName != "" {
+			{
+				var targetMS *clusterv1.MachineSet
+				for _, ms2 := range scope.machineSets {
+					if ms2.Name == targetMSName {
+						targetMS = ms2
+						break
+					}
+				}
+				if targetMS == nil {
+					log.Logf("[MS controller] - PANIC! %s is set to send replicas to %s, which does not exists", ms.Name, targetMSName)
```

Comment:
How would the test detect if we hit this line?

Reply:
I'm now returning error